### DISCUSSION TITLE
How to go about and implement both tests with external data and non-external data?

### DISCUSSION THREAD

**Discussion by YoraiLevi** [pester/Pester#2404] [General]
I'm trying to set up tests using Pester with both external and non-external data. I've created separate test files for these scenarios and used Invoke-Pester with containers, but I'm facing a couple of issues:

Split Summaries: When I run the tests with external data, the summaries are split, and it doesn't provide a clear overall picture of test results.

```
Tests Passed: 1, Failed: 1, Skipped: 0 NotRun: 0
...
Tests Passed: 1, Failed: 0, Skipped: 0 NotRun: 0
```
rather than
```
Tests Passed: 2, Failed: 1, Skipped: 0 NotRun: 0
```

Passing Configuration: I'm unsure how to pass a configuration (-Configuration) into the inner Invoke-Pester call (`Invoke-ExternalData.Tests.ps1`)

Here's the structure I'm using:

NonDataExternal.Tests.ps1:

```ps1
Describe "behaviour" {
    It "should be able to access the parameter" {
        "value" | Should -Be "value"
    }
}
```
ExternalData.ps1:

```ps1
param(
    [Parameter(Mandatory)]
    $Param
)

Describe "behaviour" {
    It "should be able to access the parameter" {
        $Param | Should -Be "value"
    }
}
```
Invoke-ExternalData.Tests.ps1:

```ps1
$values = 'value', 'not value'
$containers = $values | ForEach-Object {
    New-PesterContainer -Path "$PSScriptRoot/ExternalData.ps1" -Data @{ Param = $_ }
}
Invoke-Pester -Container $containers
```
I would appreciate guidance on how to structure these tests to achieve a unified summary and how to pass a configuration into the inner Invoke-Pester call. Is there a way to "hand-off" the PesterContainer objects without invoking Pester again?

**Comment by YoraiLevi** [comment]
My solution is to change `ExternalData.ps1` and `Invoke-ExternalData.Tests.ps1`  
`-ForEach` essentially injects the variables that are available in discovery to the Run phase
ExternalData.ps1
```ps1
param(
    [Parameter(Mandatory)]
    $Param
)
BeforeDiscovery {
    # Gather initial values for generating tests
    $ExtraParam = 'value'
}
Describe 'Test Container' -ForEach @(@{Param = $Param; ExtraParam = $ExtraParam }) {
    BeforeAll {

    }
    BeforeEach {

    }
    AfterEach {

    }
    AfterAll {

    }
    It 'Test with value <Param>' {
        $Param | Should -Not -Be $null
        $ExtraParam | Should -Not -Be $null
    }
}
```

Invoke-ExternalData.Tests.ps1
```ps1
BeforeDiscovery {
    $values = ,'value', 'not value'
}
Describe 'Test with value <value>' -ForEach $values {
    & "$PSScriptRoot/ExternalData.ps1" $_
}
```

**Reply by fflaten** [reply]
Great that you figured it out :)
The vital change here is to avoid running multiple `Invoke-Pester` (side by side or nested) which will split your results.

They way I'd approach this depends on how you get your external data.
- If it's a simple and quick lookup used in a single file: Use `BeforeDiscovery` to grab data and `-ForEach` to loop in that specific test-file. Also makes it easier to run that file only.
- If the lookup is slow, complex (requires auth etc.) or used in multiple test-files, I'd use a `test.ps1` file to configure and invoke Pester, so you pass the data to multiple files etc. E.g.

  ```powershell
  # test.ps1
  # Grab data
  $values = 'value', 'not value'
  
  # Using advanced configuration as it provides more options.
  # No downside when it's invoked through a script like this.
  $conf = New-PesterConfiguration
  
  # Simple files. Can also be added using New-PesterContainer and Run.Container below
  $conf.Run.Path = "$PSScriptRoot/NonDataExternal.Tests.ps1"
  
  # Tests with external data
  $containers = @()
  $containers += $values | ForEach-Object {
      New-PesterContainer -Path "$PSScriptRoot/ExternalData.Tests.ps1" -Data @{ Param = $_ }
      # You could also provide the whole $values array to the script and loop it using `Describe .. -ForEach $Param`
      # Just an alternative, not really better or worse. :-)
  }
  $conf.Run.Container = $containers
  
  # Other options
  $conf.Output.Verbosity = 'Detailed'
  
  Invoke-Pester -Configuration $conf
  ```

**Reply by YoraiLevi** [reply]
I want to keep allowing a naive user (aka me) to just use `Invoke-Pester` without any configurations and for things to "just work".

from my understanding, this setup which I have seen suggested on the docs, I need to introduce another script for invoking Pester. just so that `ExternalData.Tests.ps1` runs correctly. if `ExternalData.Tests.ps1` accepts mandatory parameters invoking Pester as is will prompt the console...
Is there a way to address these concerns other than what I found? or I am doing something wrong?

**Reply by fflaten** [reply]
Not really. As always it depends. If you require external data, try to fetch most of it during `BeforeDiscovery` like your `Invoke-ExternalData.Tests.ps1` example. If you require manual user input, parameters are the only way.

If you're using the same user input or external data in multiple files, I'd recommend a script like `test.ps1` to avoid duplicate download/processing and user inputs.
